transformer_lens/components.py

```diff
@@ -757,6 +789,49 @@ def apply_rotary(
 
         return torch.cat([x_rotated, x_pass], dim=-1)
 
+    def build_alibi_tensor(
+        self,
+        attention_mask: torch.Tensor,  # batch pos
+        num_heads: int,
+        dtype: torch.dtype,
+    ) -> Float[torch.Tensor, "batch head_index 1 pos"]:
+        batch_size, seq_length = attention_mask.shape
+        closest_power_of_2 = 2 ** math.floor(math.log2(num_heads))
+        base = torch.tensor(
+            2 ** (-(2 ** -(math.log2(closest_power_of_2) - 3))),
+            device=attention_mask.device,
+            dtype=torch.float32,
+        )
+        powers = torch.arange(
+            1, 1 + closest_power_of_2, device=attention_mask.device, dtype=torch.int32
+        )
+        slopes = torch.pow(base, powers)
+
+        if closest_power_of_2 != num_heads:
+            extra_base = torch.tensor(
+                2 ** (-(2 ** -(math.log2(2 * closest_power_of_2) - 3))),
+                device=attention_mask.device,
+                dtype=torch.float32,
+            )
+            num_remaining_heads = min(
+                closest_power_of_2, num_heads - closest_power_of_2
+            )
+            extra_powers = torch.arange(
+                1,
+                1 + 2 * num_remaining_heads,
+                2,
+                device=attention_mask.device,
+                dtype=torch.int32,
+            )
+            slopes = torch.cat([slopes, torch.pow(extra_base, extra_powers)], dim=0)
+
+        arange_tensor = ((attention_mask.cumsum(dim=-1) - 1) * attention_mask)[
+            :, None, :
+        ]
+        alibi = slopes[..., None] * arange_tensor
```

Comment:
Makes sense - worth adding a comment using p5 of https://arxiv.org/pdf/2108.12409.pdf to explain a
bit more about how this gets to the head-specific slope (m).

But can we use their general formula to simplify this a bit ("In general, for n heads, our set of slopes is the geometric sequence that starts
at 2(-8/n) and uses that same value as its ratio.")

Reply:
For this function, I was also using [HF's implementation](https://github.com/huggingface/transformers/blob/41496b95da5d34368a13538548c85527377972a6/src/transformers/models/bloom/modeling_bloom.py#L86). If you prefer, I can switch it back to the general (original?) formula.